Update EIP-2315: Keep on puttering (#5594)
ethereum/EIPs: EIPS/eip-2315.md

## Abstract

> This proposal provides a _complete_, _static_, _safe_ and _efficient_ control-flow facility.

Reviewer: ```suggestion
This proposal provides a _complete_, _static_, _safe_ and _efficient_ control-flow facility for EVM.
```

> It deprecates `JUMP` and `JUMPI`.

Reviewer: ```suggestion
It deprecates `JUMP` and `JUMPI` opcodes.
```

> It depends on the two opcodes proposed by [EIP-4200](./eip-4200.md):

Reviewer: ```suggestion
It depends on the two new opcodes proposed by [EIP-4200](./eip-4200.md):
```

* `RJUMP relative_offset` — relative jump to `PC + relative_offset`
* `RJUMPI relative_offset` — conditional relative jump

> It introduces two opcodes to support calling and returning from subroutines:

Reviewer: ```suggestion
It introduces two new opcodes to support calling and returning from subroutines:
```

* `RJUMPSUB relative_offset` -- relative jump to subroutine
* `RETURNSUB` -- return to `PC` after most recent `RJUMPSUB`.

> In concert with [EIP-3540](./eip-4200.md) and [EIP-3670](./eip-3670.md) it ensures, at initialization time, that valid code will not execute invalid instructions or jump to invalid locations, will not underflow stack, will maintain consistent numbers of inputs and outputs for subroutines, and will have bounded stack height in the absence of recursion.

Reviewer: ```suggestion
In concert with [EIP-4200](./eip-4200.md) and [EIP-3670](./eip-3670.md) it ensures, at initialization time, that valid code will not execute invalid instructions or jump to invalid locations, will not underflow stack, will maintain consistent numbers of inputs and outputs for subroutines, and will have bounded stack height in the absence of recursion.
```

Reviewer: EIP-3540 -> EIP-4200 to be match with filename.

Author: I think I gout the filename wrong.

Valid code will support streaming, one-pass, and other near-linear compilers.

> This is among the simplest possible proposals that meets these requirements.

Reviewer: ```suggestion
This EIP aims at achieving the simplest possible proposal among all proposals that meets these requirements.
```
Suggest change the assertive tone of judgement into either
  a subjective tone of "aims at", OR
  an objective tone of "it's simpler than XXX, YYY solutions by metrics/criteria of ZZZ"

Author: There are a few very simple ways to do it.  This is one of them.

## Motivation

### We need a complete control-flow facility.

Jumps, conditional jumps and subroutines were proposed by Alan Turing in 1945 as a means of organizing the logic of the code and the design of the memory crystals for his Automatic Computing Engine:
> "We wish to be able to arrange that sequences of orders can divide at various points, continuing in different ways according to the outcome of the calculations to date...  We also wish to be able to arrange for the splitting up of operations into subsidiary operations... To start on a subsidiary operation we need only make a note of where we left off the major operation and then apply the first instruction of the subsidiary.  When the subsidiary is over we look up the note and continue with the major operation."
> 

> > — Alan Turing — in [B.E. Carpenter, R.W. Doran, "The other Turing machine." The Computer Journal, Volume 20, Issue 3, DOI: 10.1093/comjnl/20.3.269, January 1977.](https://academic.oup.com/comjnl/article/20/3/269/751954)

Reviewer: I strongly support including this reference link

Author: Yep.  This will do until we have consensus on that.

> In more contemporary terms, we have sequences of instructions, jumps and conditional jumps that divide sequences into blocks, subroutine calls, and a stack of addresses to return to.  The details vary, but similar facilities have proven their value across a long line of important machines over the last 75 years, from the ACE to the Cray to the ARM.

Reviewer: > similar facilities have proven their value across a long line of important machines
Citation needed to backup this argument, if I may suggest...

> Unlike these machines, the Ethereum Virtual Machine _does not_ provide subroutine operations.  Instead, they must be synthesized using the dynamic `JUMP` instruction, which takes its destination on the stack. Further, the EVM provides _only_ dynamic jumps, impeding the static analysis we need.

Reviewer: ```suggestion
Unlike these machines, the Ethereum Virtual Machine _does not_ provide subroutine operations. Instead, they must be synthesized using the dynamic `JUMP` instruction, which takes its destination on the stack. Further, the EVM provides _only_ dynamic jumps, impeding the static analysis we need.
```

> ### We need safe control-flow.

Reviewer: ```suggestion
###  The control-flow needs to be safer
```

Optional, just to see if this way puts more emphasize on the "safer"

Author: I think I sort of weasel on "safe" versus "safer" throughout.  Perfect safety is impossible.

The EVM has unusually high requirements for safety.  Not only do many smart contracts control inordinately large amounts of valuable Ether, but once placed on the blockchain any defects are visible to attackers and cannot be repaired.  In valid code is an  We need to statically validate important safety constraints code at initialization time.

> ### We need static control-flow.

Reviewer: ```suggestion
### The control-flow needs to be static
```
```suggestion
###  The control-flow needs to be safer
```

Optional, just to see if this way puts more emphasize on the characteristic this EIP intend to achieve

Author: Versus keeping section titles short and avoiding passive voice.

The EVM's lone dynamic jumps cause two major problems. First, the need to synthesize static jumps and subroutines with dynamic jumps uses space and gas with needlessly complex code, as we will show below.

Worse, jumps that can dynamically branch to any destination in the code can cause quadratic "path explosions" when traversing the flow of control. For Ethereum, this is a denial-of-service vulnerability that prevents us, at initialization time, from validating the safe use of EVM code and from compiling EVM code to machine code.

> ### We need efficient control-flow.

Reviewer: ```suggestion
### The control-flow needs to be static
```
```suggestion
###  The control-flow needs to be efficient
```

Optional, just to see if this way puts more emphasize on the characteristic this EIP intend to achieve

Author: Same issue.  As section titles even the "We need" could be dropped, but it emphasizes that these are motivations.

Efficient to write by hand, compile from high level labguages, validate at deploy time, interpret by VMs, and compile to machine code.

> Static jumps, conditional jumps, and subroutines are sufficient and efficient in space and time, as shown by historical experience and as we will show for the EVM below.

Reviewer: "as shown by historical experience", it's a great argument, I'd suggest adding some reference /citation to support this.

Author: That's what the reference to Turing's is for, and a little later I list a number of important machines.  Some people don't think it's an argument at all.

**We need static jumps and subroutines and must deprecate dynamic jumps.**

> ## Specification

Reviewer: [Peer Review] I happen to notice there is no mention of "fork block number". Is this intentional? What is the plan for rolling out? Does it depend on fork block number or does it depend on total difficulty or any other approach to enable the new opcodes and deprecate the old one?

Author: It can't be known until the fork is actually scheduled, and doesn't seem to me as part of the technical specification anyway.

Reviewer: [Peer Review] I would suggest adding something like `As of FORK_BLOCK_NUMBER the implementation of EVM MUST support the following new opcodes and deprecate the following old opcodes`.

This follows many previous Core EIPs (e.g. eip-2718, eip-7, eip-140). It also bring the convenience when you need to discuss and describe scenarios of backward compatibility.

Returns control to the caller of a subroutine.
1. Pop the `return stack` to `PC`.

> The gas cost is _verylow_.

Reviewer: Is this _verylow_ referring to the yellowpaper categorization of _verrylow_? If so, can you add reference?

Author: Yellow Paper is assumed here, yes.  A few other EIPs probably need a similar reference.

Requiring a consistently aligned`data stack` also allows some algorithms that traverse the control-flow graph -- includinng code validation and compilation -- to break cycles at joins, again preventing quadratic path explosion.  When a traversal gets to a `PC` it has visited before it is either at the beginning of a loop or the entry to a function.  Since the stacks align we know that loops will not grow, or that the number of arguments is the same, so there is no need to continue that path.

_Note: The JVM and Wasm enforce similar constraints for similar reasons._

### Alternative Designs

There are a few major designs for a subroutine facility, two of which are considered here.  The others are mostly not appropriate for the EVM, such as the Wheeler Jump -- self-modifying code that writes return addresses into called subroutines.

*1. Keep return addresses on a dedicated return stack.* Turing's design is often used by stack machines, including those for Forth, Java, Wasm, and others.  The data stack is used almost entirely for computation, with a dedicated stack for return addresses.  A single instruction suffices to call, and another to return from a routine.

> *2. Keep return addresses on the data stack.*  This design is often used by register machines, including those from CDC, IBM, DEC, Intel, and ARM.  The registers are used primarily for computation, and the stack maintains call frames for return addresses, arguments, and local variables.  On the EVM there are no registers for computation, so using the stack for both purposes can be less than ideal.

Author: In earlier versions I referenced only some of the models I had actually programmed, (Burroughs 5000, CDC 7600, IBM 360, DEC PDP-11 and VAX, M68000, Sun SPARC, ARM, and Intel x86s) having been accused of "arguing from authority."  I've since decided that I am an authority, and that for a "person skilled in the art" I didn't need to be that specific.

> *2. Keep return addresses on the data stack.*  This design is often used by register machines, including those from CDC, IBM, DEC, Intel, and ARM.  The registers are used primarily for computation, and the stack maintains call frames for return addresses, arguments, and local variables.  On the EVM there are no registers for computation, so using the stack for both purposes can be less than ideal.

Reviewer: I obviously held different views with whoever made that accusation. My stance is that: so long as the EIP content is not being in-appropriately promoting, the more specific and well-referencing, the better. It helps readability.

Please feel free to ignore or make your own judgement (as the author EIP). And I'd argument _for_ author judgement.

> *2. Keep return addresses on the data stack.*  This design is often used by register machines, including those from CDC, IBM, DEC, Intel, and ARM.  The registers are used primarily for computation, and the stack maintains call frames for return addresses, arguments, and local variables.  On the EVM there are no registers for computation, so using the stack for both purposes can be less than ideal.

Author: I'm walking a Iine between keeping it brief and keeping it informative, thanks.  As the pure spec that @MicahZoltu would prefer (and I would to, in my way) I could just add a few lines to the abstract and stop there.

> *2. Keep return addresses on the data stack.*  This design is often used by register machines, including those from CDC, IBM, DEC, Intel, and ARM.  The registers are used primarily for computation, and the stack maintains call frames for return addresses, arguments, and local variables.  On the EVM there are no registers for computation, so using the stack for both purposes can be less than ideal.

Reviewer: My stance is 
1. for Spec be as brief as possible, but
2. for Motivation and Rationale, be as informative as possible. 

Since this is in the Rationale / Alternative Design, I think it's ok to be informative.
Going back to what "EIP" means to me, EIP spells Ethereum Improvement Proposal. Therefore, as a "proposal", I think it's good to elaborate reasoning as long as it's not duplicative and repetitive.
